rir/src/compiler/pir/instruction.h

```diff
@@ -209,11 +210,13 @@ class InstructionImplementation : public Instruction {
         bool LeaksEnv;
         bool HasEnv;
         bool AccessesEnv;
+        bool mayElideEnv;
```

Comment:
I don't really like this new description field. It feels kind of redundant. Can we do this instead:

```
diff --git a/rir/src/compiler/pir/instruction.h b/rir/src/compiler/pir/instruction.h
index 2864b53..c8055f0 100644
--- a/rir/src/compiler/pir/instruction.h
+++ b/rir/src/compiler/pir/instruction.h
@@ -4,6 +4,7 @@
 #include "R/r.h"
 #include "instruction_list.h"
 #include "pir.h"
+#include "env.h"
 #include "singleton_values.h"
 #include "tag.h"
 #include "value.h"
@@ -73,7 +74,6 @@ enum class EnvAccess : uint8_t {
     Read,
     Write,
     Leak,
-    NoneOrLeak,
 };
 
 // Effect that can be produced by an instruction.
@@ -208,20 +208,21 @@ class InstructionImplementation : public Instruction {
         bool HasEffect;
         bool ChangesEnv;
         bool LeaksEnv;
-        bool HasEnv;
+        bool MightHaveEnv;
         bool AccessesEnv;
-        bool mayElideEnv;
     };
 
     static constexpr InstrDescription Description = {
         EFFECT > Effect::None,    ENV >= EnvAccess::Write,
         ENV == EnvAccess::Leak,   ENV > EnvAccess::None,
-        ENV > EnvAccess::Capture, ENV >= EnvAccess::NoneOrLeak};
+        ENV > EnvAccess::Capture};
 
     bool hasEffect() const final { return Description.HasEffect; }
     bool changesEnv() const final { return Description.ChangesEnv; }
     bool leaksEnv() const final { return Description.LeaksEnv; }
-    bool hasEnv() const final { return Description.HasEnv; }
+    bool hasEnv() const final {
+        return Description.MightHaveEnv && env() != Env::elided();
+    }
     bool accessesEnv() const final { return Description.AccessesEnv; }
 
     static const Base* Cast(const Value* i) {
```

(not a complete patch)

Reply:
so basically, the static description field is `mightHaveEnv` and we use that one whenever we need to know the static size of the instruction.

The static description `hasEnv` goes away, because morally with this change, we cannot tell statically anymore.

then the dynamic `isEnv()` function will check if it's actually present.